iree/hal/cuda/cuda_driver.c

```diff
@@ -0,0 +1,210 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/base/tracing.h"
+#include "iree/hal/cuda/api.h"
+#include "iree/hal/cuda/cuda_device.h"
+#include "iree/hal/cuda/dynamic_symbols.h"
+#include "iree/hal/cuda/status_util.h"
+
+typedef struct {
+  iree_hal_resource_t resource;
+  iree_allocator_t host_allocator;
+  // Identifier used for the driver in the IREE driver registry.
+  // We allow overriding so that multiple CUDA versions can be exposed in the
+  // same process.
+  iree_string_view_t identifier;
+  int default_device_index;
+  // CUDA symbols.
+  iree_hal_cuda_dynamic_symbols_t syms;
+} iree_hal_cuda_driver_t;
+
+// Pick a fixed lenght size for device names.
+typedef char cuda_device_name_t[100];
```

Comment:
Style: it's worth keeping the array specifier on the use sites so that it's clear to readers that variables using it are fixed-length arrays (vs a cstring/string_view/etc) as memory ownership and other issues get obscured otherwise. You can define a constant length and use that (`char my_name[IREE_MAX_CUDA_DEVICE_NAME_LENGTH];`).

Reply:
Sounds good, changed it.